### ISSUE TITLE
make_sections problem

### ISSUE THREAD

**Issue by kysko** [jgm/pandoc#5846]
I've been toying a bit with the new make_sections that will appear in 2.8 (using `pandoc-windows-x86_64-2f32b5316` from 2019-10-11, no more recents nightly builds for Windows), and I've encountered what could be a problem. I've hesitated making a new issue, or posting in #5761 to which it might be related.

Minimal example: `pandoc test.txt -L test.lua --atx-headers -o test_lua.txt` with:

`test.txt`:

```
::: {.mydiv}
# header 1a

one

# header 1b

two
:::
```

and

`test.lua`:

```
function Pandoc(doc)
 local b = pandoc.utils.make_sections(true, 1, doc.blocks)
 return pandoc.Pandoc(b, doc.meta)
end
```

which give the result:

`test_lua.txt`:

```
::: {#header-1a .section .mydiv}
# header 1a {number="1"}

one
:::

::: {#header-1b .section}
# header 1b {number="2"}

two
:::
```
Notice how `.mydiv` now affects only the first header, not the second as originally intended.

I was expecting something like:

```
::: {.mydiv}
::: {#header-1a .section}
# header 1a {number="1"}

one
:::

::: {#header-1b .section}
# header 1b {number="2"}

two
:::
:::
```

or maybe at least:

```
::: {#header-1a .section .mydiv}
# header 1a {number="1"}

one
:::

::: {#header-1b .section .mydiv}
# header 1b {number="2"}

two
:::
```

(What would be desirable might be related to  #5761)

In any case, the result is problematic. Imagine for example `.mydiv` was to exclude those sections. Of course one could rewrite the original div and enclose each in `.mydiv`... but still, this should be documented (as a *caveat* or something) if it's currently the expected result.

**Comment by jgm** [comment]
I think this is the same issue discussed over in #5761, but specialized to lua.
So probably this should be closed so we can have discussion in one place.
The lua function uses the same underlying code as `--section-divs`.

**Comment by kysko** [comment]
ok, wanted to make sure. 

`--section-divs` didn't show any `<section>` tags or `:::` markup for `markdown` output, so I figured it was only for `HTML`.

Anyways, my surrounding divs where mainly for (lua) manipulations, and `make_sections` could just be enough. I just need to know clearly the impact on what I've already marked up.

so, closing.

**Comment by jgm** [comment]
I want to keep this open a bit longer while I think about it.

**Comment by kysko** [comment]
ok.
I was going to post this (partly) in #5761, so here goes:

Inheritance can be achieved, if one wants to, through a filter (propagated with walkblock). As you said in #5671, keep things simple. One can then complexify at will. So, in my case above, I would expect the first one, where the original div isn't touched.

I understand there may be more annoying cases to parse, a simple example being:

```
::: mydiv
# head 1
## head 1.1
# head 2
:::

## head 2.1
```
I'd never do that, but if it occurs, currently this becomes (with my above pandoc cmd line):

```
::: {#head-1 .section .mydiv}
# head 1 {number="1"}

::: {#head-1.1 .section}
## head 1.1 {number="1.1"}
:::
:::

::: {#head-2 .section}
# head 2 {number="2"}
:::

::: {#head-2.1 .section}
# head 2.1 {number="3"}
:::
```
and what I would expect is something like:

```
::: {.mydiv}
::: {#head-1 .section}
# head 1 {number="1"}

::: {#head-1.1 .section}
## head 1.1 {number="1.1"}
:::
:::

::: {#head-2 .section}
# head 2 {number="2"}
:::
:::

::: {#head-2.1 .section}
# head 2.1 {number="3"}
:::
```
So basically the same thing, but with the `mydiv` respected.
If I want the mydiv class to propagate to one or all of the inside divs, I could code it myself.

(`## head 2.1` becomes level 1 because where my `mydiv` ended, which is *my* problem in the first place, not the `make_section` algorithm.)

Don't know the implications for the *big picture* though.

**Comment by jgm** [comment]
By the way @kysko, you can get nightlies (for windows too) from https://github.com/jgm/pandoc/actions

**Comment by kysko** [comment]
oh, ok, thanks!
was getting them from pandoc-extras...

**Comment by kysko** [comment]
Hmmm... just noticed this:
By simply inserting a comment after the opening div:
```
::: mydiv
<!-- -->
# head 1
## head 1.1
# head 2
:::

## head 2.1
```
then the output is as expected.
```
::: {.mydiv}
<!-- -->

::: {#head-1 .section}
# head 1 {number="1"}

::: {#head-1.1 .section}
## head 1.1 {number="1.1"}
:::
:::

::: {#head-2 .section}
# head 2 {number="2"}
:::
:::

::: {#head-2.1 .section}
# head 2.1 {number="3"}
:::
```
(Inserting blank lines won't work though)
So `make_sections` doesn't really have a *fundamental* problem with this, but has a glitch when the header is preceded by an opening Div.
This doesn't solve the technical question in #5761, but it mitigates the (or *my*) problem in the current situation.

**Comment by jgm** [comment]
The code is pattern matching for a Div whose contents start with a Header; that's why inserting a comment changes things.

I think one easy change would be to check to see if the Div contains multiple top-level Headers; if so, maybe it should just remain as it is and we should create new Divs for the sections within it.

**Comment by kysko** [comment]
Makes sense to me.

Although there will be pathological cases (as always 😄  ).
For example, a problematic case presented by brainchild in #5761:

```
# One
## One A
::: fence
## One B
# Two
:::
```
This would force `make_sections` not to respect the level of `# Two`, and the original output is:
```
::: {#one .section}
# One {number="1"}

::: {#one-a .section}
## One A {number="1.1"}
:::

::: {#one-b .section .fence}
## One B {number="1.2"}
:::

::: {#two .section}
## Two {number="1.3"}
:::
:::
```
The modified output by inserting the comment line is:
```
::: {#one .section}
# One {number="1"}

::: {#one-a .section}
## One A {number="1.1"}

::: {.fence}
<!-- -->

::: {#one-b .section}
### One B {number="1.1.1"}
:::

::: {#two .section}
### Two {number="1.1.2"}
:::
:::
:::
:::
```
In both cases `make_sections` recognized the div.fence as a block within `One A` header 'content'. The difference is the header levels created within the block after.

I did expect level 2 headers for both `One B` and `Two`, and not level 3.

The user should beware these pathological situations, just as, say, embedded emphasis like `*one *two* three*`. Too easy to create complex and ambiguous situations. Can't please everyone.

In the case above, *IF* `div.fence` represented some state to affect its content blocks independently (e.g. to be hidden, or to be extracted, or other *distributive* behavior), maybe the user could have used:
```
# One
## One A
::: fence
## One B
:::
::: fence
# Two
:::
```
in which case `make_sections` gives:
```
::: {#one .section}
# One {number="1"}

::: {#one-a .section}
## One A {number="1.1"}
:::

::: {#one-b .section .fence}
## One B {number="1.2"}
:::
:::

::: {#two .section .fence}
# Two {number="2"}
:::
```
... which looks much better. But that will depend on the nature of the `div.fence` and the user.
